### PR TITLE
Setting default values for Deployment

### PR DESCRIPTION
**Pull request by nikhiljindal**
Ref: #1743

While making the change, also realized a few issues in the API objects:
- Changed uniqueLabelKey to be a string field rather than a pointer in the internal API.
- Changed maxSurge and maxUnavailable to be pointers, so that they can be defaulted to 1.

cc @wojtek-t @bgrant0607 @ironcladlou @ghodss @caesarxuchao

### CODE REVIEW

File: pkg/expapi/types.go [hunk 1]
```diff
@@ -199,7 +199,7 @@ type DeploymentSpec struct {
 	// not add any selector and label. If unspecified, system uses
 	// "deployment.kubernetes.io/podTemplateHash".
 	// Value of this key is hash of DeploymentSpec.PodTemplateSpec.
-	UniqueLabelKey *string `json:"uniqueLabel,omitempty"`
+	UniqueLabelKey string `json:"uniqueLabel,omitempty"`
```
Comment (smarterclayton) [inline pkg/expapi/types.go]: Does the JSON intentionally not match?

Comment (thockin) [inline pkg/expapi/types.go]: +1

**Comment by thockin** [comment]
mostly lgtm

File: pkg/expapi/types.go (hunk 1 above)
Comment (nikhiljindal) [inline pkg/expapi/types.go]: Fixed

**Comment by nikhiljindal** [comment]
Thanks for pointing that out. Have fixed code as per comment.

**Comment by nikhiljindal** [comment]
@k8s-bot ok to test

**Comment by smarterclayton** [comment]
Something I thought of while looking at this - we have no way to add custom
deployment strategies on top of deployments, and we have no easy path
forward for users / customers to extend their own types.  We don't have
"preserve unknown JSON" options at this time, but I'd like to make a
determination about how we want to treat strategies for deployment going
forward (can I have two strategies at the same time, like canary and
rolling, or are they exclusive and someone has to implement each type).

On Wed, Sep 2, 2015 at 3:57 AM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *passed* for commit 30a5832
> <https://github.com/kubernetes/kubernetes/commit/30a58321e8c17f2f7d59c2787260c4ffe147e080>
> .
>
>    - Build Log
>    <https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/30a58321e8c17f2f7d59c2787260c4ffe147e080/kubernetes-pull-build-test-e2e-gce/7100/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/30a58321e8c17f2f7d59c2787260c4ffe147e080/kubernetes-pull-build-test-e2e-gce/7100/_artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//7100>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13489#issuecomment-136968840>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by ironcladlou** [comment]
This LGTM

**Pull request merged**